types/src/contract_messages.rs

```diff
@@ -17,77 +29,195 @@ use rand::{
 #[cfg_attr(feature = "datasize", derive(DataSize))]
-pub struct MessageTopicAddr {
+pub struct MessageAddr {
     /// The entity addr.
     entity_addr: HashAddr,
     /// The hash of the name of the message topic.
-    topic_hash: MessageTopicHash,
+    topic_name_hash: TopicNameHash,
+    /// The message index.
+    message_index: Option<u32>,
 }
 
-impl MessageTopicAddr {
-    /// Constructs a new [`MessageAddr`] based on the addressable entity addr and the hash of the
+impl MessageAddr {
+    /// Constructs a new topic address based on the addressable entity addr and the hash of the
     /// message topic name.
-    pub const fn new(entity_addr: HashAddr, topic_hash: MessageTopicHash) -> Self {
+    pub const fn new_topic_addr(entity_addr: HashAddr, topic_name_hash: TopicNameHash) -> Self {
         Self {
             entity_addr,
-            topic_hash,
+            topic_name_hash,
+            message_index: None,
+        }
+    }
+
+    /// Constructs a new message address based on the addressable entity addr, the hash of the
+    /// message topic name and the message index in the topic.
+    pub const fn new_message_addr(
+        entity_addr: HashAddr,
+        topic_name_hash: TopicNameHash,
+        message_index: u32,
+    ) -> Self {
+        Self {
+            entity_addr,
+            topic_name_hash,
+            message_index: Some(message_index),
+        }
+    }
+
+    /// Formats the [`MessageAddr`] as a prefixed, hex-encoded string.
+    pub fn to_formatted_string(self) -> String {
+        match self.message_index {
+            Some(index) => {
+                format!(
+                    "{}{}-{}-{:x}",
+                    MESSAGE_ADDR_PREFIX,
+                    base16::encode_lower(&self.entity_addr),
+                    self.topic_name_hash.to_formatted_string(),
+                    index,
+                )
+            }
+            None => {
+                format!(
+                    "{}{}{}-{}",
+                    MESSAGE_ADDR_PREFIX,
+                    TOPIC_FORMATTED_STRING_PREFIX,
+                    base16::encode_lower(&self.entity_addr),
+                    self.topic_name_hash.to_formatted_string(),
+                )
+            }
+        }
+    }
+
+    /// Parses a formatted string into a [`MessageAddr`].
+    pub fn from_formatted_str(input: &str) -> Result<Self, FromStrError> {
+        let remainder = input
+            .strip_prefix(MESSAGE_ADDR_PREFIX)
+            .ok_or(FromStrError::InvalidPrefix)?;
+
+        let (remainder, message_index) = match remainder.strip_prefix(TOPIC_FORMATTED_STRING_PREFIX)
+        {
+            Some(topic_string) => (topic_string, None),
+            None => {
+                let parts = input.rsplitn(2, '-').collect::<Vec<_>>();
+                if parts.len() != 2 {
+                    return Err(FromStrError::MissingMessageIndex);
+                }
+                (parts[1], Some(u32::from_str_radix(parts[0], 16)?))
```

Comment:
how about
```suggestion
                let (remainder, message_index_str) = input.rsplit_once('-').ok_or(FromStrError::MissingMessageIndex)?;
                let message_index = u32::from_str_radix(message_index_str, 16)?;
```
same can be done to simplify other usages of `splitn(2, *)` and `rsplitn(2, *)` as well

Reply:
changed here: https://github.com/casper-network/casper-node/pull/4315/commits/566b3534a29355dd56ec8375f4c42d0a2ca0f24c